### PR TITLE
Move static Jekyll posts to dynamic Vue rendered posts served over Express.

### PR DESCRIPTION
**Pull request by 3vorp**
Our current post system is one of the big things preventing a rewrite from going well, since it's so coupled to the design of the site and Jekyll. To mitigate this, I've completely moved posts from Jekyll to being served with Express and rendered with Vue (much like add-ons). In addition to being a lot easier to deal with, this means we can move posts over to the API now and have the Jekyll site still function.

Posts still look pretty much the same, but are internally much cleaner (the current post system uses JS string interpolation to build parts of the HTML document :skull:). There's also decent embed and SEO support (which is why I used express and not just pure Vue).

Quick note—I had to move the posts folder because Jekyll refused to ignore it otherwise; there's no other reason.

### CODE REVIEW

File: js/posts/post-changelog.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+export default {
+	name: "post-changelog",
+	// recursive component (best way I could find to do this)
```
Comment (TheRolfFR) [inline js/posts/post-changelog.js:3]: okay, the code is clean so it's good

File: post.html [hunk 2]
```diff
@@ -0,0 +1,13 @@
+---
+layout: default
+title: "%title%"
+description: "%description%"
+url: "%permalink%"
+header-img: "%header_img%"
+---
+<script>
+window.postData = `%postData%`;
```
Comment (TheRolfFR) [inline post.html:9]: instead of putting a string here, and parse it in the Vue, can't we juste put the JSON directly here?

File: news.html [hunk 3]
```diff
@@ -5,23 +5,41 @@
...
-      <img
-        src="https://database.faithfulpack.net/images/website/posts/placeholder.jpg"
-        loading="lazy"
-        alt="{{ post.title }}"
-      >
-      {%- endif -%}
-      <h3>{{ post.title }}</h3>
-    </a>
-  {%- endfor -%}
+    <!-- need to use v-text because jekyll parses the braces otherwise -->
```
Comment (TheRolfFR) [inline news.html:21]: even with backslashes ?

File: js/posts/index.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+				</template>
+				<br />
+				<a href="https://discord.gg/sN9YRQbBv7">
+					<p class="blurple banner">Start a discussion in our Discord!</p>
+				</a>
+			</div>
+		`,
+		data() {
+			return {
+				data: JSON.parse(window.postData),
```
Comment (TheRolfFR) [inline js/posts/index.js:38]: See parse comment below

**Review by TheRolfFR — CHANGES_REQUESTED** [review]
Good overall

File: index.html [hunk 5]
```diff
@@ -140,16 +140,18 @@ <h3>See More</h3>
...
+      v-for="post in topPosts"
+      class="card img-card"
+      rel="noopener"
+      :href="'{{ site.baseurl }}' + post.permalink"
+    >
+      <img v-if="post.header_img" :src="post.header_img" loading="lazy">
+      <img v-else src="https://database.faithfulpack.net/images/website/posts/placeholder.jpg" loading="lazy">
+      <!-- need to use v-text because jekyll parses the braces otherwise -->
+      <h3 v-text="post.title"></h3>
+    </a>
```
[the remark this answers is not in the document]
Comment (3vorp) [reply index.html:154]: we aren't using vuetify

File: faq.html [hunk 6]
```diff
@@ -6,7 +6,7 @@
 <link href="https://cdn.jsdelivr.net/npm/vuetify@2.6.3/dist/vuetify.min.css" rel="stylesheet">
 <div id="app"></div>
 
-<script defer src="js/axios.min.js"></script>
+<script defer src="{{ site.baseurl }}/js/axios.min.js"></script>
 <script defer src="https://cdn.jsdelivr.net/npm/marked@4.1.0/marked.min.js"></script>
```
[the remark this answers is not in the document]
Comment (3vorp) [reply faq.html:10]: fantastic question, it's been like that always for some reason

File: index.js [hunk 7]
```diff
@@ -59,50 +63,46 @@ function walkSync(dir, filelist = []) {
...
-				delete obj.layout;
-				const [y, m, d, ...name] = filename.split("-");
-				obj.date = [y, m, d].join("-");
-				const key = name.join("-").replace(".md", "");
-				if (acc[key]) reject(new Error(`Duplicate post key ${key}`));
-				acc[key] = obj;
-				return acc;
-			}, {});
+		const result = paths.reduce((acc, cur) => {
+			const filename = cur.split("/").at(-1);
```
[the remark this answers is not in the document]
Comment (3vorp) [reply index.js:72]: good idea

File: post.html (hunk 2 above)
Comment (3vorp) [reply post.html:9]: Technically yes but you'd have to work around liquid with interpolations (like on the news page) and I found using string templates much cleaner

File: news.html (hunk 3 above)
Comment (3vorp) [reply news.html:21]: then vue wouldn't detect it lol

File: index.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (3vorp) [reply index.html]: it's at the very bottom of the page so I figured that by the time you scroll to it it would have loaded

**Review by TheRolfFR — CHANGES_REQUESTED** [review]
Last edits and you can merge

File: index.html [hunk 8]
```diff
@@ -120,18 +120,18 @@ <h1 class="display-3" align="center">Add-ons</h1>
...
         <div class="img-card-shadow"></div>
         <h3 v-text="addon.name"></h3>
         <div class="addon-flags" style="margin-bottom: 5px">
-          <img v-show="addon.options.tags.includes('Java')" style="margin-bottom: 5px" src="/image/icon/java.png" alt="available for Java Edition" loading="lazy">
-          <img v-show="addon.options.tags.includes('Bedrock')" style="margin-bottom: 5px" src="/image/icon/bedrock.png" alt="available for Bedrock Edition" loading="lazy">
-          <img v-show="addon.options.optifine" style="margin-bottom: 5px" class="optifine" src="/image/icon/optifine.png" alt="requires optifine" loading="lazy">
+          <img v-show="addon.options.tags.includes('Java')" style="margin-bottom: 5px" src="image/icon/java.png" alt="available for Java Edition" loading="lazy">
+          <img v-show="addon.options.tags.includes('Bedrock')" style="margin-bottom: 5px" src="image/icon/bedrock.png" alt="available for Bedrock Edition" loading="lazy">
+          <img v-show="addon.options.optifine" style="margin-bottom: 5px" class="optifine" src="image/icon/optifine.png" alt="requires optifine" loading="lazy">
```
[the remark this answers is not in the document]
Comment (3vorp) [reply index.html:129]: whoops I think my regex messed up a bit

File: backend/post.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+router.get("/posts.json", cors(corsOptions), async (_, res) => {
+	const posts = await getPostJSON();
+	if (!posts) return res.status(400).send({ message: "Could not load posts!" }).end();
+	return res.send(JSON.stringify(posts)).end();
+});
+
+// match all routes and continue if no matching post is found
+router.get("*", async (req, res, next) => {
+	const posts = await getPostJSON();
```
[the remark this answers is not in the document]
Comment (3vorp) [reply backend/post.js:89]: ooh that's smart, I forgot that since this is esm we can use top level await

File: backend/post.js [hunk 10]
```diff
@@ -71,45 +71,61 @@ async function getPostJSON() {
...
+		)
+		.replaceAll("%postData%", cleanPostData(post));
+	return data;
+}
+
+// force new generation on startup (prevents new routes not being added)
+const posts = await generatePostJSON();
+
+// match all post routes
+router.get(Object.keys(posts), async (req, res) => {
```
Comment (TheRolfFR) [inline backend/post.js:101]:
```suggestion
// include the ending slash URLs here
const post_urls = Object.keys(posts).map(url => [url, url + '/']).flat();
// to match them here
router.get(post_urls, async (req, res) => {
        // then delete the slash back
```

File: js/addons/addon-post.js [hunk 11]
```diff
@@ -232,13 +222,24 @@ export default {
...
+    header() {
+      return this.files.find((el) => el.use === 'header').source
+    },
+    carousel() {
+      return this.files.filter((el) => el.use === 'carousel' || el.use === 'screenshot').map((el) => el.source)
+    },
+    downloads() {
+      return this.files.filter((el) => el.use === 'download')
+    }
+  },
```
Comment (TheRolfFR) [inline js/addons/addon-post.js:235]: good idea

**Review by TheRolfFR — CHANGES_REQUESTED** [review]
lastest tiny change and you can merge without review

File: backend/post.js (hunk 10 above)
Comment (3vorp) [reply backend/post.js:101]: isn't needed, the last slash gets matched by express either way (it just shows up in the query params and the post keys don't include it so we need to remove it)

**Pull request merged**